### PR TITLE
External interrupt

### PR DESCRIPTION
**Pull request by connojd**
Add support for trapping on external interrupts (vectors 32-255). Integration tests will be added later with the interrupt management code

### CODE REVIEW

File: include/hve/arch/intel_x64/external_interrupt.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+private:
+
+    /// @cond
+
+    void enable_exiting();
+
+    /// @endcond
+
+    std::array<std::list<handler_delegate_t>, 256> m_handlers{};
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/external_interrupt.h]: remove {}

File: include/hve/arch/intel_x64/external_interrupt.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    std::array<uint64_t, 256> m_log{};
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/external_interrupt.h]: remove {}

File: src/hve/arch/intel_x64/external_interrupt.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+{
+    if (!ndebug && m_log_enabled) {
+        dump_log();
+    }
+}
+
+void
+external_interrupt::add_handler(
+    vmcs_n::value_type vector, handler_delegate_t &&d)
+{ m_handlers[vector].push_front(std::move(d)); }
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/external_interrupt.cpp]: This is unprotected. Use .at() instead.

File: src/hve/arch/intel_x64/external_interrupt.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+// Basic exit reason handler
+// -----------------------------------------------------------------------------
+
+bool
+external_interrupt::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    info_t info{vmcs_n::vm_exit_interruption_information::vector::get()};
+
+    if (!ndebug && m_log_enabled) {
+        m_log[info.vector]++;
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/external_interrupt.cpp]: use .at()

File: src/hve/arch/intel_x64/external_interrupt.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// -----------------------------------------------------------------------------
+// Basic exit reason handler
+// -----------------------------------------------------------------------------
+
+bool
+external_interrupt::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    info_t info{vmcs_n::vm_exit_interruption_information::vector::get()};
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/external_interrupt.cpp]: Just to stay consistent:

```
struct info_t info = {
    vmcs_n::vm_exit_interruption_information::vector::get()
};
```

File: src/hve/arch/intel_x64/external_interrupt.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        m_log[info.vector]++;
+    }
+
+    for (const auto &d : m_handlers[info.vector]) {
+        if (d(vmcs, info)) {
+            return true;
+        }
+    }
+
+    return false;
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/external_interrupt.cpp]: Returning false here is a bad idea as that will result in an unhandled VM exit error as the base hypervisor doesn't support this type of exit. Likely this should just throw a useful error here.

**Comment by connojd** [comment]
@rianquinn good to go now

**Pull request merged**